src/main/java/net/fabricmc/tinyremapper/FileSystemReference.java

```diff
@@ -0,0 +1,150 @@
+/**
+ * Reference counted fs handling.
+ *
+ * <p>The implementation closes file systems opened by itself once they are closed as often as they were opened. This
+ * allows intersecting open+close actions on e.g. the zip based file systems. The caller has to ensure open and close
+ * invocations are mirrored.
+ */
+public final class FileSystemReference implements Closeable {
+	public static FileSystemReference openJar(Path path) throws IOException {
+		return openJar(path, false);
+	}
+
+	public static FileSystemReference openJar(Path path, boolean create) throws IOException {
+		return open(toJarUri(path), create);
+	}
+
+	private static URI toJarUri(Path path) {
+		URI uri = path.toUri();
+
+		try {
+			return new URI("jar:" + uri.getScheme(), uri.getHost(), uri.getPath(), uri.getFragment());
+		} catch (URISyntaxException e) {
+			throw new RuntimeException("can't convert path "+path+" to uri", e);
+		}
+	}
+
+	public static FileSystemReference open(URI uri) throws IOException {
+		return open(uri, false);
+	}
+
+	public static FileSystemReference open(URI uri, boolean create) throws IOException {
+		synchronized (openFsMap) {
+			boolean opened = false;
+			FileSystem fs = null;
+
+			try {
+				fs = FileSystems.getFileSystem(uri);
+			} catch (FileSystemNotFoundException e) {
+				try {
+					fs = FileSystems.newFileSystem(uri, create ? Collections.singletonMap("create", "true") : Collections.emptyMap());
+					opened = true;
+				} catch (FileSystemAlreadyExistsException f) {
+					fs = FileSystems.getFileSystem(uri);
+				}
+			}
+
+			FileSystemReference ret = new FileSystemReference(fs);
+			Set<FileSystemReference> refs = openFsMap.get(fs);
+
+			if (refs == null) {
+				refs = Collections.newSetFromMap(new IdentityHashMap<>());
+				openFsMap.put(fs, refs);
+				if (!opened) refs.add(null);
+			} else if (opened) {
+				throw new IllegalStateException("opened but already in refs?");
+			}
+
+			refs.add(ret);
+
+			return ret;
+		}
+	}
+
+	private FileSystemReference(FileSystem fs) {
+		this.fileSystem = fs;
+	}
+
+	public boolean isReadOnly() {
+		if (closed) throw new IllegalStateException("fs closed");
+
+		return fileSystem.isReadOnly();
+	}
+
+	public Path getPath(String first, String... more) {
+		if (closed) throw new IllegalStateException("fs closed");
+
+		return fileSystem.getPath(first, more);
+	}
+
+	public FileSystem getFs() {
+		if (closed) throw new IllegalStateException("fs closed");
+
+		return fileSystem;
+	}
+
+	@Override
+	public void close() throws IOException {
+		synchronized (openFsMap) {
+			if (closed) return;
+			closed = true;
+
+			Set<FileSystemReference> refs = openFsMap.get(fileSystem);
+			if (refs == null || !refs.remove(this)) throw new IllegalStateException("fs "+fileSystem+" was already closed");
+
+			if (refs.isEmpty()) {
+				openFsMap.remove(fileSystem);
+				fileSystem.close();
+			} else if (refs.size() == 1 && refs.iterator().next() == null) { // only null -> not opened by us, just abandon
```

Comment:
```suggestion
			} else if (refs.size() == 1 && refs.contains(null)) { // only null -> not opened by us, just abandon
```
Avoids allocation of iterator; and since identity hash map is a sparse hash table, iteration may be slower than hash lookup. See https://github.com/openjdk/jdk/blob/17cc7131abf699070dc491e69b21207f01c65cf6/src/java.base/share/classes/java/util/IdentityHashMap.java#L356